src/engine/LocalVocab.cpp

```diff
@@ -57,6 +57,7 @@ LocalVocabIndex LocalVocab::getIndexAndAddIfNotContainedImpl(WordT&& word) {
       wordsToIndexesMap_.insert({std::forward<WordT>(word), nextFreeIndex_});
   const auto& [wordInMap, index] = *wordInMapAndIndex;
   if (isNewWord) {
+    AD_CONTRACT_CHECK(!readOnly_);
```

Comment:
I would move the check right to the beginning of the function, 
I think it is already an error to even call this function on a `readOnly_` vocabulary (For `const` access there is `getWord` if I remember correctly).